### PR TITLE
[BugFix] Lockfree enqueue check if it is cancelled

### PR DESCRIPTION
**Pull request by liuyehcf**
## Background

https://github.com/StarRocks/starrocks/pull/27515 this pr remove the lock around the enqueu process, and it only double-checks the short circuit but misses the `_is_cancelled`, so there's still a chance that a closure may be added into the queue, during which the query has been cancelled, and finally resulted in the closue leak, and RPC timeout.

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

**Review by silverbullet233 — DISMISSED** [review]

File: be/src/runtime/sender_queue.cpp [hunk 1]
```diff
@@ -775,7 +770,7 @@ Status DataStreamRecvr::PipelineSenderQueue::add_chunks(const PTransmitChunkPara
             _chunk_queue_states[index].blocked_closure_num += closure != nullptr;
             _total_chunks++;
             // Double check here for short circuit compatibility without introducing a critical section
-            if (_chunk_queue_states[index].is_short_circuited.load(std::memory_order_relaxed)) {
+            if (_is_cancelled || _chunk_queue_states[index].is_short_circuited.load(std::memory_order_relaxed)) {
                 short_circuit(index);
```
Comment (fzhedu) [inline be/src/runtime/sender_queue.cpp:774]: no need break?

Comment (liuyehcf) [reply be/src/runtime/sender_queue.cpp:774]: Yes, the logic is shared by short circuit and cancel. Short circuit may occur for some of parallelism, and the other parallelism need to proceed.

File: be/src/runtime/sender_queue.cpp [hunk 2]
```diff
@@ -790,20 +793,18 @@ void DataStreamRecvr::PipelineSenderQueue::short_circuit(const int32_t driver_se
...
+    while (chunk_queue.size_approx() > 0) {
+        if (chunk_queue.try_dequeue(item)) {
+            if (item.closure != nullptr) {
+                COUNTER_UPDATE(metrics.closure_block_timer, MonotonicNanos() - item.queue_enter_time);
+                item.closure->Run();
+                chunk_queue_state.blocked_closure_num--;
             }
+            --_total_chunks;
+            _recvr->_num_buffered_bytes -= item.chunk_bytes;
+            COUNTER_ADD(metrics.peak_buffer_mem_bytes, item.chunk_bytes);
```
Comment (silverbullet233) [inline be/src/runtime/sender_queue.cpp:807]: there are same logics in clear_buffer_queues and short_circuit, could we add a new function and put it into the function?e.g. clear_chunk_queue(ChunkQueue&)

File: be/src/runtime/sender_queue.cpp [hunk 3]
```diff
@@ -539,10 +539,8 @@ void DataStreamRecvr::PipelineSenderQueue::decrement_senders(int be_number) {
...
 void DataStreamRecvr::PipelineSenderQueue::cancel() {
-    bool expected = false;
-    if (_is_cancelled.compare_exchange_strong(expected, true)) {
-        clean_buffer_queues();
-    }
+    _is_cancelled = true;
+    clean_buffer_queues();
 }
 
 void DataStreamRecvr::PipelineSenderQueue::close() {
```
Comment (fzhedu) [inline be/src/runtime/sender_queue.cpp:546]: it seems strange that normal close call cancel

File: be/src/runtime/sender_queue.cpp (hunk 2 above)
Comment (liuyehcf) [reply be/src/runtime/sender_queue.cpp]: The short circuit is only used for pipeline level shuffle and it should be kept. And I've reverted the previous change, PTAL again.

File: be/src/runtime/sender_queue.cpp (hunk 3 above)
Comment (liuyehcf) [reply be/src/runtime/sender_queue.cpp:546]: updated.

**Review by fzhedu — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

**Pull request merged**